src/servers/src/prom.rs

```diff
@@ -463,6 +477,208 @@ pub async fn range_query(
+            .collect()
+    });
+    let form_matches: Option<Vec<String>> = form_params.matches.0.map(|s| {
+        s.split('$')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect()
+    });
+
+    if matches.is_none() && form_matches.is_none() {
+        return PromJsonResponse::error("Unsupported", "match[] parameter is required");
+    }
+
+    let queries = matches.or(form_matches).unwrap();
+
+    let start = params
+        .start
+        .or(form_params.start)
+        .unwrap_or_else(day_before_current_time_rfc3339);
+    let end = params
+        .end
+        .or(form_params.end)
+        .unwrap_or_else(current_time_rfc3339);
+
+    let db = &params.db.unwrap_or(DEFAULT_SCHEMA_NAME.to_string());
+    let (catalog, schema) = super::parse_catalog_and_schema_from_client_database_name(db);
+
+    let mut labels: HashSet<String> = HashSet::new();
+    labels.insert(METRIC_NAME.to_string());
+
+    for query in queries {
+        let prom_query = PromQuery {
+            query,
+            start: start.clone(),
+            end: end.clone(),
+            step: DEFAULT_LOOKBACK_STRING.to_string(),
+        };
+
+        let query_ctx = QueryContext::with(catalog, schema);
+        let result = handler.do_query(&prom_query, Arc::new(query_ctx)).await;
+
+        let response = retrieve_labels_name_from_query_result(result, &mut labels).await;
+
+        if let Err(err) = response {
+            // Prometheus won't report error if querying nonexist label and metric
+            if err.status_code() != StatusCode::TableNotFound
+                && err.status_code() != StatusCode::TableColumnNotFound
+            {
+                return PromJsonResponse::error(err.status_code().to_string(), err.to_string());
+            }
+        }
+    }
+
+    let mut sorted_labels: Vec<String> = labels.into_iter().collect();
+    sorted_labels.sort();
+    PromJsonResponse::success(PromData::LabelsPromData(sorted_labels))
+}
+
+/// Retrieve labels name from query result
+async fn retrieve_labels_name_from_query_result(
+    result: Result<Output>,
+    labels: &mut HashSet<String>,
+) -> Result<()> {
+    match result? {
+        Output::RecordBatches(batches) => {
+            record_batches_to_labels_name(batches, labels)?;
+            Ok(())
+        }
+        Output::Stream(stream) => {
+            let batches = RecordBatches::try_collect(stream)
+                .await
+                .context(CollectRecordbatchSnafu)?;
+            record_batches_to_labels_name(batches, labels)?;
+            Ok(())
+        }
+        Output::AffectedRows(_) => Err(Error::UnexpectedResult {
+            reason: "expected data result, but got affected rows".to_string(),
+        }),
+    }
+}
+
+/// Retrieve labels name from record batches
+fn record_batches_to_labels_name(
+    batches: RecordBatches,
+    labels: &mut HashSet<String>,
+) -> Result<()> {
+    let mut tag_column_indices = Vec::new();
```

Comment:
I think we can return all the column names as labels except `greptime_value`. It's not precise, but acceptable.

The `greptime_value` is defined at https://github.com/GreptimeTeam/greptimedb/blob/497b1f9dc987e118bbd482e0b92e36889a33db7c/src/servers/src/prometheus.rs#L35

Reply:
Do your mean we return all the column names including timestamp column name and field column name.